src/phoenix/core/model_schema.py

```diff
@@ -0,0 +1,1067 @@
+
+
+class TimeRange(NamedTuple):
+    start: datetime
+    stop: datetime
+
+
+class ModelData(ObjectProxy, ABC):  # type: ignore
+    def __init__(
+        self,
+        data: DataFrameOrSeries,
+        /,
+        _model: Optional["ProxyType[Model]"] = None,
+    ) -> None:
+        super().__init__(data)
+        self._self_model = _model
+
+    @property
+    @abstractmethod
+    def null_value(self) -> Any:
+        ...
+
+    def __getitem__(self, key: Any) -> Any:
+        if _is_column_key(key):
+            if isinstance(key, Column):
+                return key(self)
+            if self._self_model is None or self.empty:
+                return self.null_value
+            model = cast(Model, self._self_model)
+            return model[key](self)
+        return super().__getitem__(key)
+
+
+class EventId(NamedTuple):
+    """Identifies an event."""
+
+    row_id: int = 0
+    dataset_id: DatasetRole = PRIMARY
+
+    def __str__(self) -> str:
+        return ":".join(map(str, self))
+
+
+class Event(ModelData):
+    def __init__(
+        self,
+        series: "pd.Series[Any]",
+        /,
+        event_id: EventId,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(series, **kwargs)
+        self._self_id = event_id
+
+    @property
+    def id(self) -> EventId:
+        return self._self_id
+
+    def null_value(self) -> float:
+        return float("nan")
+
+    @overload
+    def __getitem__(self, key: ColumnKey) -> Any:
+        ...
+
+    @overload
+    def __getitem__(self, key: Any) -> Any:
+        ...
+
+    def __getitem__(self, key: Any) -> Any:
+        return super().__getitem__(key)
+
+
+class Events(ModelData):
+    def __init__(
+        self,
+        df: pd.DataFrame,
+        /,
+        role: DatasetRole,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(df, **kwargs)
+        self._self_role = role
+
+    @property
+    def null_value(self) -> Any:
+        return pd.Series(dtype=object)
+
+    @property
+    def time_range(self) -> TimeRange:
+        if self._self_model is None or self.empty:
+            # NOTE: as of Python 3.8.16, pandas 1.5.3:
+            # >>> isinstance(pd.NaT, datetime.datetime)
+            # True
+            return TimeRange(pd.NaT, pd.NaT)  # type: ignore
+        return self._time_range
+
+    @cached_property
+    def _time_range(self) -> TimeRange:
+        model = cast(Model, self._self_model)
+        min_max = _agg_min_max(model[TIME](self))
+        start_time = cast(datetime, min_max.min())
+        end_time = cast(datetime, min_max.max())
+        # Add one minute to end_time, because time intervals are right
+        # open and one minute is the smallest interval allowed.
+        stop_time = end_time + timedelta(minutes=1)
+        # Round down to the nearest minute.
+        start = start_time.replace(second=0, microsecond=0)
+        stop = stop_time.replace(second=0, microsecond=0)
+        return TimeRange(start, stop)
```

Comment:
this indirection doesn't make too much sense to me. Why is not just in the above?